asab/storage/mongodb.py

```diff
@@ -28,6 +30,63 @@
 )
 
 
+def transactional(method_to_decorate):
+	"""
+	A decorator for asynchronous methods that wraps execution within a MongoDB transaction.
+
+	Ensures ACID properties for operations. Handles session management, transaction
+	start/commit/abort, and specific error conditions like `DryRunAbort` and `DuplicateError`.
+	"""
+	@functools.wraps(method_to_decorate)
+	async def wrapper(*args, **kwargs):
+		inst_self = args[0]
+		decorated_method_args = args[1:]
+
+		if isinstance(inst_self, asab.storage.mongodb.MongoDBUpsertor):
+			client = inst_self.Storage.Client
+			obj_id = inst_self.ObjId
+		else:
+			client = inst_self.Client
+			_, obj_id = decorated_method_args
+
+		async with await client.start_session() as session:
+			try:
+				session.start_transaction(
+					read_concern=pymongo.read_concern.ReadConcern("local"),
+					write_concern=pymongo.write_concern.WriteConcern("majority")
+				)
+				result = await method_to_decorate(inst_self, session, *decorated_method_args, **kwargs)
+
+				dry_run = DRY_RUN.get("dry_run")
+				if dry_run:
+					if session.in_transaction:
+						await session.abort_transaction()
+
+
+					raise DryRunAbort(obj_id=obj_id)
+
+				if session.in_transaction:
+					await session.commit_transaction()
+				return result
+
+			except DryRunAbort:
+				return obj_id
+			except DuplicateError:
+				if session.in_transaction:
+					await session.abort_transaction()
+				raise
```

Comment:
i suggest raising *something*. ideally 
```python
except DuplicateError as e:
				if session.in_transaction:
					await session.abort_transaction()
				raise e
```
same for the generic exception below.